### PR TITLE
fix: connection leak when initConnections

### PR DESCRIPTION
**Pull request by mikawudi**
add initing status in entry, and set entry freezed status null when all connection inited

issue: https://github.com/redisson/redisson/issues/3270

Signed-off-by: xujie <mikawudi@qq.com>

### CODE REVIEW

**Comment by mikawudi** [comment]
not wait pubSubConnectionPool inited, beacuse i think pubSubConnectionPool minimumIdleSize is alway less than 10

File: redisson/src/main/java/org/redisson/connection/balancer/LoadBalancerManager.java [hunk 1]
```diff
@@ -138,12 +138,18 @@ public boolean unfreeze(ClientConnectionsEntry entry, FreezeReason freezeReason)
...
-                entry.resetFirstFail();
-                entry.setFreezeReason(null);
-                
-                slaveConnectionPool.initConnections(entry);
-                pubSubConnectionPool.initConnections(entry);
-                return true;
+                if (!entry.isIniting()) {
+                    entry.setIniting(true);
+                    entry.resetFirstFail();
+                    slaveConnectionPool.initConnections(entry).onComplete((r, ex) -> {
```
Comment (mrniko) [inline redisson/src/main/java/org/redisson/connection/balancer/LoadBalancerManager.java:144]: please use AsyncCountDownLatch here. Both handlers slaveConnectionPool, pubSubConnectionPool invoke countDown method and in latch callback we do rest logic.

File: redisson/src/main/java/org/redisson/connection/balancer/LoadBalancerManager.java [hunk 2]
```diff
@@ -138,12 +138,18 @@ public boolean unfreeze(ClientConnectionsEntry entry, FreezeReason freezeReason)
 
             if (freezeReason != FreezeReason.RECONNECT
                     || entry.getFreezeReason() == FreezeReason.RECONNECT) {
-                entry.resetFirstFail();
-                entry.setFreezeReason(null);
-                
-                slaveConnectionPool.initConnections(entry);
-                pubSubConnectionPool.initConnections(entry);
-                return true;
+                if (!entry.isIniting()) {
```
Comment (mrniko) [inline redisson/src/main/java/org/redisson/connection/balancer/LoadBalancerManager.java:141]: use AtomicBoolean.compareAndSet method

**Review by mrniko — CHANGES_REQUESTED** [review]

File: redisson/src/main/java/org/redisson/connection/ClientConnectionsEntry.java [hunk 3]
```diff
@@ -56,6 +56,8 @@
 
     private final AtomicLong firstFailTime = new AtomicLong(0);
 
+    private volatile boolean initing = false;
```
Comment (mrniko) [inline redisson/src/main/java/org/redisson/connection/ClientConnectionsEntry.java:59]: rename this setting to `initialized`

**Review by mrniko — CHANGES_REQUESTED** [review]

File: redisson/src/main/java/org/redisson/connection/balancer/LoadBalancerManager.java [hunk 4]
```diff
@@ -138,12 +140,29 @@ public boolean unfreeze(ClientConnectionsEntry entry, FreezeReason freezeReason)
...
                     || entry.getFreezeReason() == FreezeReason.RECONNECT) {
-                entry.resetFirstFail();
-                entry.setFreezeReason(null);
-                
-                slaveConnectionPool.initConnections(entry);
-                pubSubConnectionPool.initConnections(entry);
-                return true;
+                if (!entry.isInitialized()) {
+                    entry.setInitialized(true);
+                    AtomicInteger initedCounter = new AtomicInteger(2);
```
Comment (mrniko) [inline redisson/src/main/java/org/redisson/connection/balancer/LoadBalancerManager.java:145]: we can use CountableListener here:
```java
RPromise<Void> promise = new RedissonPromise<Void>();
promise.onComplete((r, ex) -> {
     if (ex == null) {
         entry.setFreezeReason(null);
     }
});
CountableListener listener = new CountableListener(promise, null, 2);
slaveConnectionPool.initConnections(entry).onComplete(() -> {listener.decCounter()});
pubSubConnectionPool.initConnections(entry).onComplete(() -> {listener.decCounter()});
```

**Review by mrniko — CHANGES_REQUESTED** [review]

**Review by mrniko — APPROVED** [review]

**Comment by mrniko** [comment]
Thanks for contribution!

**Pull request merged**